File: src/telemetry/status.cpp
#include "telemetry/status.h"

#include <google/protobuf/util/json_util.h>
#include <tuple>

namespace Serializers
{
namespace Telemetry
{
MessagesPerSecond::MessagesPerSecond(const PbTelemetry::MessagesPerSecond& protobuf) {
    deviceName = protobuf.devicename();
    bitsPerSecond = protobuf.bitspersecond();
    busLoad = protobuf.busload();
    count = protobuf.count();
}

MessagesPerSecond::operator PbTelemetry::MessagesPerSecond() const {
    PbTelemetry::MessagesPerSecond ret;
    ret.set_devicename(deviceName);
    ret.set_bitspersecond(bitsPerSecond);
    ret.set_busload(busLoad);
    ret.set_count(count);
    return ret;
}

std::string MessagesPerSecond::serializeAsJsonString() const {
    PbTelemetry::MessagesPerSecond protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string MessagesPerSecond::serializeAsProtobufString() const {
    PbTelemetry::MessagesPerSecond protobuf(*this);
    return protobuf.SerializeAsString();
}

bool MessagesPerSecond::deserializeFromJsonString(const std::string& str) {
    PbTelemetry::MessagesPerSecond protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool MessagesPerSecond::deserializeFromProtobufString(const std::string& str) {
    PbTelemetry::MessagesPerSecond protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

Status::Status(const PbTelemetry::Status& protobuf) {
    timestamp = protobuf.timestamp();
    zeroTimestamp = protobuf.zerotimestamp();
    state = static_cast<TelemetryState>(protobuf.state());
    cpuTotalLoad = protobuf.cputotalload();
    cpuProcessLoad = protobuf.cpuprocessload();
    memProcessLoad = protobuf.memprocessload();
    canlibBuildTime = protobuf.canlibbuildtime();
    telemetryBuildTime = protobuf.telemetrybuildtime();
    messagesPerSecond = {protobuf.messagespersecond().begin(), protobuf.messagespersecond().end()};
}

Status::operator PbTelemetry::Status() const {
    PbTelemetry::Status ret;
    ret.set_timestamp(timestamp);
    ret.set_zerotimestamp(zeroTimestamp);
    ret.set_state(static_cast<PbTelemetry::TelemetryState>(state));
    ret.set_cputotalload(cpuTotalLoad);
    ret.set_cpuprocessload(cpuProcessLoad);
    ret.set_memprocessload(memProcessLoad);
    ret.set_canlibbuildtime(canlibBuildTime);
    ret.set_telemetrybuildtime(telemetryBuildTime);
    *(ret.mutable_messagespersecond()) = {messagesPerSecond.begin(), messagesPerSecond.end()};
    return ret;
}

std::string Status::serializeAsJsonString() const {
    PbTelemetry::Status protobuf(*this);
    std::string ret;
    google::protobuf::util::JsonPrintOptions options;
    options.add_whitespace = true;
    std::ignore = google::protobuf::util::MessageToJsonString(protobuf, &ret, options);
    return ret;
}

std::string Status::serializeAsProtobufString() const {
    PbTelemetry::Status protobuf(*this);
    return protobuf.SerializeAsString();
}

bool Status::deserializeFromJsonString(const std::string& str) {
    PbTelemetry::Status protobuf;
    auto status = google::protobuf::util::JsonStringToMessage(str, &protobuf);
    if(status.ok()) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}

bool Status::deserializeFromProtobufString(const std::string& str) {
    PbTelemetry::Status protobuf;
    if(protobuf.ParseFromString(str)) {
        *this = protobuf;
        return true;
    } else {
        return false;
    }
}
}
}

File: py/telemetry/status.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict

from telemetry import status_pb2
from google.protobuf.json_format import MessageToJson, Parse

class TelemetryState(Enum):
    INIT = 0
    IDLE = 1
    SETUP_RUN = 2
    RUN = 3
    WAIT_READY = 4
    ERROR = 5


@dataclass
class MessagesPerSecond:
    deviceName: str = ""
    bitsPerSecond: int = 0
    busLoad: float = 0.0
    count: int = 0
    
    _proto_message: status_pb2.MessagesPerSecond = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = status_pb2.MessagesPerSecond()

    def _populate_proto(self):
        self._proto_message.deviceName = self.deviceName
        self._proto_message.bitsPerSecond = self.bitsPerSecond
        self._proto_message.busLoad = self.busLoad
        self._proto_message.count = self.count

    @classmethod
    def _from_proto(cls, proto_message) -> "MessagesPerSecond":
        return cls(
            deviceName = proto_message.deviceName,
            bitsPerSecond = proto_message.bitsPerSecond,
            busLoad = proto_message.busLoad,
            count = proto_message.count,
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "MessagesPerSecond":
        message = status_pb2.MessagesPerSecond()
        message.ParseFromString(data)
        return cls(
            deviceName = message.deviceName,
            bitsPerSecond = message.bitsPerSecond,
            busLoad = message.busLoad,
            count = message.count,
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "MessagesPerSecond":
        message = status_pb2.MessagesPerSecond()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())

@dataclass
class Status:
    timestamp: int = 0
    zeroTimestamp: int = 0
    state: TelemetryState = None
    cpuTotalLoad: int = 0
    cpuProcessLoad: int = 0
    memProcessLoad: int = 0
    canlibBuildTime: int = 0
    telemetryBuildTime: int = 0
    messagesPerSecond: List[MessagesPerSecond] = field(default_factory=list)
    
    _proto_message: status_pb2.Status = field(init=False, repr=False)

    def __post_init__(self):
        self._proto_message = status_pb2.Status()

    def _populate_proto(self):
        self._proto_message.timestamp = self.timestamp
        self._proto_message.zeroTimestamp = self.zeroTimestamp
        self._proto_message.state = self.state.value
        self._proto_message.cpuTotalLoad = self.cpuTotalLoad
        self._proto_message.cpuProcessLoad = self.cpuProcessLoad
        self._proto_message.memProcessLoad = self.memProcessLoad
        self._proto_message.canlibBuildTime = self.canlibBuildTime
        self._proto_message.telemetryBuildTime = self.telemetryBuildTime
        del self._proto_message.messagesPerSecond[:]
        for val in self.messagesPerSecond:
            val._populate_proto()
            tmp = self._proto_message.messagesPerSecond.add()
            tmp.CopyFrom(val._proto_message)

    @classmethod
    def _from_proto(cls, proto_message) -> "Status":
        return cls(
            timestamp = proto_message.timestamp,
            zeroTimestamp = proto_message.zeroTimestamp,
            state = TelemetryState._from_proto(proto_message.state),
            cpuTotalLoad = proto_message.cpuTotalLoad,
            cpuProcessLoad = proto_message.cpuProcessLoad,
            memProcessLoad = proto_message.memProcessLoad,
            canlibBuildTime = proto_message.canlibBuildTime,
            telemetryBuildTime = proto_message.telemetryBuildTime,
            messagesPerSecond=[MessagesPerSecond._from_proto(val) for val in proto_message.messagesPerSecond],
        )

    def __str__(self):
        return self.serializeAsJsonString()

    def serializeAsProtobufString(self) -> bytes:
        self._populate_proto()
        return self._proto_message.SerializeToString()

    @classmethod
    def deserializeFromProtobufString(cls, data: bytes) -> "Status":
        message = status_pb2.Status()
        message.ParseFromString(data)
        return cls(
            timestamp = message.timestamp,
            zeroTimestamp = message.zeroTimestamp,
            state = TelemetryState(message.state),
            cpuTotalLoad = message.cpuTotalLoad,
            cpuProcessLoad = message.cpuProcessLoad,
            memProcessLoad = message.memProcessLoad,
            canlibBuildTime = message.canlibBuildTime,
            telemetryBuildTime = message.telemetryBuildTime,
            messagesPerSecond = [MessagesPerSecond._from_proto(val) for val in message.messagesPerSecond],
        )

    def serializeAsJsonString(self) -> str:
        self._populate_proto()
        return MessageToJson(self._proto_message)

    @classmethod
    def deserializeFromJsonString(cls, data: str) -> "Status":
        message = status_pb2.Status()
        Parse(data, message)
        return cls.deserializeFromProtobufString(message.SerializeToString())
